Parca: Update supported version to v0.19+

metalmatze opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Parca has moved on to using Apache Arrow to represent the flame graph. 
It's a lot more performant than the previous API response. 

**Why do we need this feature?**

The Parca team hasn't developed the old API any further and there is no plan to do anything with it. 
Additionally, the new API is a lot faster.

**Who is this feature for?**

This is important for all Parca users as the API is a lot more stable and performant than the old API. 
The Parca users will get faster responses and updates to the newer API. 

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #59834

**Special notes for your reviewer:**

Please check that:
- [*] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

metalmatze committed:
pkg/tsdb/parca: Upgrade to using the flamegraph arrow

metalmatze committed:
pkg/tsdb/parca: Delete code for old flamegraphs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=89826) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=89826) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=89826) it.</sub>

metalmatze committed:
pkg/tsdb/parca: Fix golangci-lint error in test

metalmatze commented:
Any first impressions @gabor @yesoreyeram @ivanahuckova?

ivanahuckova commented:
@grafana/observability-traces-and-profiling @aocenas would you have a capacity with helping us with reviewing this one?

aocenas commented on pkg/tsdb/parca/query.go:
```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
 	} else {
-		panic("unknown report type returned from query")
+		// TODO: Can we be nicer about signaling users to update to have the latest APIs?
+		panic("unknown report type returned from query. update parca?")
```
Is there a way to support both? I assume in the future they can diverge so we should push people to use the new API but I would think for now, supporting both and logging some warning would be nicer to the users.

metalmatze replied:
We don't have a meta API to discover which APIs are supported. Maybe gRPC has something via reflection. Or we simply try the newer one and fall back to the older one if it is not available. 
In reality, I don't expect many people to encounter this. There are so many developments in Parca that most people quickly upgrade.

aocenas replied:
Hmm I would not mind deferring to you on this but guess @ivanahuckova should have the final say as they are the owners of this now.

In case we go with just hard update, I guess it would make more sense to then instead of panicking return the error in the response. Before the panic was there because I didn't think that could reasonably happen but in this case it would be sort of expected sometimes.

ivanahuckova replied:
I agree with @aocenas, I would prefer if we supported both, but if you @metalmatze think that the effort does not make sense, then I would:
- Instead of panicking, returning Error repose that would mention min parca version that user should update to
- Adding supported parca version to the documentation (see example in Loki data source documentation: https://grafana.com/docs/grafana/latest/datasources/loki/#supported-loki-versions)
- Adding information that the supported version of parca was changed to changelog

metalmatze replied:
Added the note to the Parca datasource documentation. 
I'm not sure how to add this change to the changelog. Do I simply add a line to `CHANGELOG.md`? I'm not sure where exactly this would fit? Which version etc.

ivanahuckova replied:
Done ✅  - I've added  `add to changelog` label and updated the PR title that will be in the changelog.

metalmatze replied:
Updated. 
This somehow got lost during a rebase, I think.

aocenas commented on pkg/tsdb/parca/query.go:
```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
+	arrowReader, err := ipc.NewReader(bytes.NewBuffer(flamegraph.GetRecord()))
+	if err != nil {
+		// TODO: Handle properly?
+		return nil
```
Can we just return it? I mean before we didn't return `(dataFrame, error)` response but mainly because there was nothing in the processing that would return error that we couldn't handle in place, here though seems like more things can go wrong so I would return it and pass into the `response.Error` in the `query()` function.

metalmatze replied:
Yes, happy to return errors. I just thought it was best practice not to return errors here in these plugins.

aocenas replied:
I guess we are not always super consistent with this but errors should be returned upstream and let either front-end or some middleware deal with it, if they cannot be recovered in the plugin.

metalmatze replied:
Fixed.

aocenas commented:
Added some comments but need to take a closer look on Monday.

aocenas commented on pkg/tsdb/parca/query.go:
```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
+	defer arrowReader.Release()
+
+	arrowReader.Next()
+	rec := arrowReader.Record()
```
Bit unclear to me here, that we get a `record`, then put it in the table just so we can do `rec := tr.Record()` first thing in the `newFlamegraphIterator` func. Is that necessary?

metalmatze replied:
Let me check again, but I'm certain that the record reader has some sanity checks underneath.

metalmatze replied:
Removed it. We don't need this wrapper. We can directly read the record as we will only ever have one record here.

aocenas commented on pkg/tsdb/parca/query.go:
```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
+	rec := tr.Record()
+	schema := rec.Schema()
+
+	columnChildren := rec.Column(schema.FieldIndices(FlamegraphFieldChildren)[0]).(*array.List)
```
Is there some doc about the arrow columns structure, how the flame graph is encoded in it? Didn't work with arrow before and even though I think I know how arrow works I am confused about what the columns mean in context of the flamegraph data.

metalmatze replied:
Sadly not. 
It's super close to the frames you use in your backend.
Each column is an array that encodes all values. 
To read a row you need to find the value in all arrays at the given index.

The children column is a list of lists. You can think of it as this: 
```json
[[0, 1, 2, 3], [4, 5], [6, 7], null, ...]
````
It's always indices pointing to other rows.
The first row is always our root row in the flame graph. Then this root row points to `[0, 1,2, 3]` in the example above. Once we've read the first row we know we need to walk these other rows to find the children of a row. 

I hope that explains the children column a bit better.

aocenas replied:
Ah ok yeah now it makes more sense, thanks.

metalmatze committed:
pkg/tsdb/parca: Handle errors nicely

metalmatze committed:
docs/sources/datasource: Add Parca v0.19+ support note

metalmatze committed:
pkg/tsdb/parca: Don't use arrow table reader

As pointed out during reviews, it's not really needed and we can read the record directly.

metalmatze commented:
Addressed most comments.
Thank you for your patience. 
I'm a bit swamped with other Polar Signals things and organizing PromCon on the side.

aocenas commented:
I think there is still the one panic call that we may want to change but otherwise I think the code is good. @ivanahuckova can I leave testing this to you?

metalmatze commented:
Let me check that panic quickly and then fix the linter issues.

metalmatze commented:
Drone says: 
> Build is blocked, please, contact repo admin in order to proceed

I guess I can't do anything about that.

ivanahuckova commented:
> I guess I can't do anything about that.
I've approved the build

aocenas commented:
@metalmatze thanks a lot for the PR. As is probably apparent we don't have enough capacity for every data source so any contribution is appreciated.

metalmatze commented:
Hey, @aocenas, no worries. 
We're happy to keep the Parca integration up-to-date whenever something changes like this time around. 
Feel free to reach out if you ever have any questions about Parca (data source) in the future.
